File: ios/Runner/LlmPlugin.swift
import Flutter
import Foundation

/// Native LLM inference plugin using llama.cpp for iOS.
///
/// Uses llama.cpp C API directly via the bridging header.
/// Metal GPU acceleration is enabled when available.
///
class LlmPlugin: NSObject {

    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    // Model state (opaque pointers to llama.cpp structs)
    private var model: OpaquePointer?
    private var ctx: OpaquePointer?

    private var isGenerating = false
    private var shouldStop = false

    private let inferenceQueue = DispatchQueue(
        label: "com.citecoach.llm.inference",
        qos: .userInitiated
    )

    func register(with messenger: FlutterBinaryMessenger) {
        methodChannel = FlutterMethodChannel(
            name: "com.citecoach/llm",
            binaryMessenger: messenger
        )
        methodChannel?.setMethodCallHandler(handle)

        eventChannel = FlutterEventChannel(
            name: "com.citecoach/llm_stream",
            binaryMessenger: messenger
        )
        eventChannel?.setStreamHandler(StreamHandler(plugin: self))
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "loadModel":
            let args = call.arguments as? [String: Any]
            let modelPath = args?["modelPath"] as? String
            loadModel(path: modelPath, result: result)

        case "startGeneration":
            let args = call.arguments as? [String: Any] ?? [:]
            let prompt = args["prompt"] as? String ?? ""
            let maxTokens = args["maxTokens"] as? Int ?? 512
            let temperature = args["temperature"] as? Double ?? 0.7
            let topP = args["topP"] as? Double ?? 0.9
            let repeatPenalty = args["repeatPenalty"] as? Double ?? 1.1
            startGeneration(
                prompt: prompt,
                maxTokens: maxTokens,
                temperature: Float(temperature),
                topP: Float(topP),
                repeatPenalty: Float(repeatPenalty),
                result: result
            )

        case "stopGeneration":
            shouldStop = true
            result(nil)

        case "unloadModel":
            unloadModel(result: result)

        case "getModelInfo":
            let info: [String: Any] = [
                "name": "Gemma 2B Instruct",
                "quantization": "Q4_K_M",
                "parameters": "2B",
                "isLoaded": model != nil
            ]
            result(info)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Model Loading

    private func loadModel(path: String?, result: @escaping FlutterResult) {
        inferenceQueue.async { [weak self] in
            guard let self = self else { return }

            let modelPath = path ?? self.getDefaultModelPath()

            guard FileManager.default.fileExists(atPath: modelPath) else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "MODEL_NOT_FOUND",
                        message: "Model file not found at: \(modelPath)",
                        details: nil
                    ))
                }
                return
            }

            // Initialize llama.cpp backend
            llama_backend_init()

            // Configure model parameters with Metal GPU acceleration
            var modelParams = llama_model_default_params()
            modelParams.n_gpu_layers = 99 // Offload all layers to Metal GPU

            // Load the model
            guard let loadedModel = llama_model_load_from_file(modelPath, modelParams) else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "LOAD_FAILED",
                        message: "Failed to load model from: \(modelPath)",
                        details: nil
                    ))
                }
                return
            }
            self.model = loadedModel

            // Create inference context
            var ctxParams = llama_context_default_params()
            ctxParams.n_ctx = 2048
            ctxParams.n_threads = UInt32(min(ProcessInfo.processInfo.activeProcessorCount, 4))
            ctxParams.n_batch = 512

            guard let context = llama_init_from_model(loadedModel, ctxParams) else {
                llama_model_free(loadedModel)
                self.model = nil
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "CONTEXT_FAILED",
                        message: "Failed to create inference context",
                        details: nil
                    ))
                }
                return
            }
            self.ctx = context

            DispatchQueue.main.async {
                result(true)
            }
        }
    }

    // MARK: - Text Generation

    private func startGeneration(
        prompt: String,
        maxTokens: Int,
        temperature: Float,
        topP: Float,
        repeatPenalty: Float,
        result: @escaping FlutterResult
    ) {
        guard let model = self.model, let ctx = self.ctx else {
            result(FlutterError(
                code: "MODEL_NOT_LOADED",
                message: "Model is not loaded",
                details: nil
            ))
            return
        }

        guard !isGenerating else {
            result(FlutterError(
                code: "ALREADY_GENERATING",
                message: "Generation already in progress",
                details: nil
            ))
            return
        }

        isGenerating = true
        shouldStop = false
        result(nil) // Acknowledge start, tokens come via EventChannel

        inferenceQueue.async { [weak self] in
            guard let self = self else { return }
            defer { self.isGenerating = false }

            let vocab = llama_model_get_vocab(model)

            // Tokenize the prompt
            let promptCStr = prompt.cString(using: .utf8)!
            let nPromptMax = Int32(prompt.utf8.count + 256)
            var tokens = [llama_token](repeating: 0, count: Int(nPromptMax))
            let nTokens = llama_tokenize(vocab, promptCStr, Int32(promptCStr.count - 1),
                                          &tokens, nPromptMax, true, true)

            if nTokens < 0 {
                DispatchQueue.main.async {
                    self.eventSink?("[ERROR]Failed to tokenize prompt")
                }
                return
            }
            tokens = Array(tokens.prefix(Int(nTokens)))

            // Clear KV cache
            llama_kv_cache_clear(ctx)

            // Process prompt in batches
            var batch = llama_batch_init(512, 0, 1)

            for i in 0..<Int(nTokens) {
                llama_batch_add(&batch, tokens[i], Int32(i), [0], false)
                if batch.n_tokens >= 512 || i == Int(nTokens) - 1 {
                    if i == Int(nTokens) - 1 {
                        batch.logits[Int(batch.n_tokens) - 1] = 1 // true
                    }
                    if llama_decode(ctx, batch) != 0 {
                        DispatchQueue.main.async {
                            self.eventSink?("[ERROR]Failed to decode prompt")
                        }
                        llama_batch_free(batch)
                        return
                    }
                    llama_batch_clear(&batch)
                }
            }

            // Set up sampler chain
            let smpl = llama_sampler_chain_init(llama_sampler_chain_default_params())!
            llama_sampler_chain_add(smpl, llama_sampler_init_temp(temperature))
            llama_sampler_chain_add(smpl, llama_sampler_init_top_p(topP, 1))
            llama_sampler_chain_add(smpl, llama_sampler_init_penalties(
                64, repeatPenalty, 0.0, 0.0))
            llama_sampler_chain_add(smpl, llama_sampler_init_dist(42))

            // Generate tokens
            var nGenerated = 0
            var nPos = Int32(nTokens)

            while nGenerated < maxTokens {
                if self.shouldStop { break }

                // Sample next token
                let newToken = llama_sampler_sample(smpl, ctx, -1)

                // Check for end of generation
                if llama_vocab_is_eog(vocab, newToken) {
                    break
                }

                // Convert token to text
                var buf = [CChar](repeating: 0, count: 256)
                let n = llama_token_to_piece(vocab, newToken, &buf, 256, 0, true)
                if n < 0 {
                    break
                }

                let tokenText = String(cString: buf)

                // Send token to Flutter via EventChannel
                DispatchQueue.main.async {
                    self.eventSink?(tokenText)
                }

                // Prepare next batch
                llama_batch_clear(&batch)
                llama_batch_add(&batch, newToken, nPos, [0], true)
                nPos += 1

                if llama_decode(ctx, batch) != 0 {
                    DispatchQueue.main.async {
                        self.eventSink?("[ERROR]Failed to decode token")
                    }
                    break
                }

                nGenerated += 1
            }

            llama_batch_free(batch)
            llama_sampler_free(smpl)

            // Signal completion
            DispatchQueue.main.async {
                self.eventSink?("[DONE]")
            }
        }
    }

    // MARK: - Model Unloading

    private func unloadModel(result: @escaping FlutterResult) {
        inferenceQueue.async { [weak self] in
            guard let self = self else { return }

            if let ctx = self.ctx {
                llama_free(ctx)
            }
            if let model = self.model {
                llama_model_free(model)
                llama_backend_free()
            }

            self.ctx = nil
            self.model = nil

            DispatchQueue.main.async {
                result(true)
            }
        }
    }

    // MARK: - Helpers

    private func getDefaultModelPath() -> String {
        let documentsDir = FileManager.default.urls(
            for: .documentDirectory,
            in: .userDomainMask
        ).first!
        return documentsDir.appendingPathComponent(
            "models/gemma-2b-it-q4"
        ).path
    }

    func dispose() {
        shouldStop = true
        inferenceQueue.async { [weak self] in
            if let ctx = self?.ctx { llama_free(ctx) }
            if let model = self?.model {
                llama_model_free(model)
                llama_backend_free()
            }
            self?.ctx = nil
            self?.model = nil
        }
        methodChannel?.setMethodCallHandler(nil)
    }

    // MARK: - Stream Handler

    class StreamHandler: NSObject, FlutterStreamHandler {
        weak var plugin: LlmPlugin?

        init(plugin: LlmPlugin) {
            self.plugin = plugin
        }

        func onListen(
            withArguments arguments: Any?,
            eventSink events: @escaping FlutterEventSink
        ) -> FlutterError? {
            plugin?.eventSink = events
            return nil
        }

        func onCancel(withArguments arguments: Any?) -> FlutterError? {
            plugin?.eventSink = nil
            return nil
        }
    }
}

File: android/app/src/main/kotlin/com/citecoach/citecoach/LlmPlugin.kt
package com.citecoach.citecoach

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Native LLM inference plugin using llama.cpp.
 *
 * This plugin bridges Flutter to the native llama.cpp library for
 * on-device LLM inference. It handles:
 * - Model loading/unloading
 * - Text generation with streaming
 * - Memory management
 *
 * ## Setup Instructions:
 *
 * 1. Add llama.cpp as a native dependency:
 *    - Download llama.cpp source from https://github.com/ggerganov/llama.cpp
 *    - Place in android/app/src/main/cpp/llama.cpp/
 *    - Or use the pre-built AAR from llama.cpp releases
 *
 * 2. Add to android/app/build.gradle:
 *    ```
 *    android {
 *        externalNativeBuild {
 *            cmake {
 *                path "src/main/cpp/CMakeLists.txt"
 *            }
 *        }
 *    }
 *    ```
 *
 * 3. Create CMakeLists.txt in android/app/src/main/cpp/:
 *    ```cmake
 *    cmake_minimum_required(VERSION 3.22)
 *    project(citecoach_llm)
 *    add_subdirectory(llama.cpp)
 *    add_library(citecoach_llm SHARED llm_bridge.cpp)
 *    target_link_libraries(citecoach_llm llama common)
 *    ```
 */
class LlmPlugin : MethodChannel.MethodCallHandler {

    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private var context: Context? = null

    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())

    // Native model handle (0 = not loaded)
    private var modelHandle: Long = 0
    private var contextHandle: Long = 0

    @Volatile
    private var isGenerating = false

    @Volatile
    private var shouldStop = false

    fun register(engine: FlutterEngine, context: Context) {
        this.context = context

        methodChannel = MethodChannel(engine.dartExecutor.binaryMessenger, "com.citecoach/llm")
        methodChannel?.setMethodCallHandler(this)

        eventChannel = EventChannel(engine.dartExecutor.binaryMessenger, "com.citecoach/llm_stream")
        eventChannel?.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "loadModel" -> {
                val modelPath = call.argument<String>("modelPath")
                loadModel(modelPath, result)
            }
            "startGeneration" -> {
                val prompt = call.argument<String>("prompt") ?: ""
                val maxTokens = call.argument<Int>("maxTokens") ?: 512
                val temperature = call.argument<Double>("temperature") ?: 0.7
                val topP = call.argument<Double>("topP") ?: 0.9
                val repeatPenalty = call.argument<Double>("repeatPenalty") ?: 1.1
                startGeneration(prompt, maxTokens, temperature.toFloat(),
                    topP.toFloat(), repeatPenalty.toFloat(), result)
            }
            "stopGeneration" -> {
                stopGeneration(result)
            }
            "unloadModel" -> {
                unloadModel(result)
            }
            "getModelInfo" -> {
                getModelInfo(result)
            }
            else -> result.notImplemented()
        }
    }

    private fun loadModel(modelPath: String?, result: MethodChannel.Result) {
        executor.execute {
            try {
                val path = modelPath ?: getDefaultModelPath()
                val file = File(path)

                if (!file.exists()) {
                    mainHandler.post {
                        result.error("MODEL_NOT_FOUND",
                            "Model file not found at: $path", null)
                    }
                    return@execute
                }

                // Load native library
                System.loadLibrary("citecoach_llm")

                // Load model via JNI
                modelHandle = nativeLoadModel(path)

                if (modelHandle == 0L) {
                    mainHandler.post {
                        result.error("LOAD_FAILED",
                            "Failed to load model", null)
                    }
                    return@execute
                }

                // Create inference context (1024 tokens - balanced for mobile)
                contextHandle = nativeCreateContext(modelHandle, 1024)

                mainHandler.post { result.success(true) }
            } catch (e: UnsatisfiedLinkError) {
                mainHandler.post {
                    result.error("NATIVE_LIB_MISSING",
                        "Native llama.cpp library not found. " +
                        "Please follow the setup instructions in LlmPlugin.kt " +
                        "to build the native library.", null)
                }
            } catch (e: Exception) {
                mainHandler.post {
                    result.error("LOAD_ERROR",
                        "Error loading model: ${e.message}", null)
                }
            }
        }
    }

    private fun startGeneration(
        prompt: String,
        maxTokens: Int,
        temperature: Float,
        topP: Float,
        repeatPenalty: Float,
        result: MethodChannel.Result
    ) {
        if (modelHandle == 0L || contextHandle == 0L) {
            result.error("MODEL_NOT_LOADED", "Model is not loaded", null)
            return
        }

        if (isGenerating) {
            result.error("ALREADY_GENERATING",
                "Generation already in progress", null)
            return
        }

        isGenerating = true
        shouldStop = false
        result.success(null) // Acknowledge start, tokens come via EventChannel

        executor.execute {
            try {
                nativeGenerate(
                    contextHandle,
                    prompt,
                    maxTokens,
                    temperature,
                    topP,
                    repeatPenalty
                ) { token ->
                    if (shouldStop) return@nativeGenerate false

                    mainHandler.post {
                        eventSink?.success(token)
                    }
                    true // Continue generating
                }

                mainHandler.post {
                    eventSink?.success("[DONE]")
                }
            } catch (e: Exception) {
                mainHandler.post {
                    eventSink?.success("[ERROR]${e.message}")
                }
            } finally {
                isGenerating = false
            }
        }
    }

    private fun stopGeneration(result: MethodChannel.Result) {
        shouldStop = true
        result.success(null)
    }

    private fun unloadModel(result: MethodChannel.Result) {
        executor.execute {
            try {
                if (contextHandle != 0L) {
                    nativeFreeContext(contextHandle)
                    contextHandle = 0
                }
                if (modelHandle != 0L) {
                    nativeFreeModel(modelHandle)
                    modelHandle = 0
                }
                mainHandler.post { result.success(true) }
            } catch (e: Exception) {
                mainHandler.post {
                    result.error("UNLOAD_ERROR",
                        "Error unloading model: ${e.message}", null)
                }
            }
        }
    }

    private fun getModelInfo(result: MethodChannel.Result) {
        val info = HashMap<String, Any>()
        info["name"] = "Qwen 2.5 1.5B Instruct"
        info["parameters"] = "1.5B"
        info["isLoaded"] = modelHandle != 0L
        result.success(info)
    }

    private fun getDefaultModelPath(): String {
        val appDir = context?.filesDir?.absolutePath ?: ""
        return "$appDir/models/qwen2.5-1.5b-instruct-q4_k_m.gguf"
    }

    fun dispose() {
        shouldStop = true
        executor.execute {
            if (contextHandle != 0L) {
                nativeFreeContext(contextHandle)
                contextHandle = 0
            }
            if (modelHandle != 0L) {
                nativeFreeModel(modelHandle)
                modelHandle = 0
            }
        }
        executor.shutdown()
        methodChannel?.setMethodCallHandler(null)
    }

    // ---- JNI Native Methods ----
    // These are implemented in the native C++ bridge (llm_bridge.cpp)
    // that links against llama.cpp

    private external fun nativeLoadModel(modelPath: String): Long
    private external fun nativeCreateContext(modelHandle: Long, contextSize: Int): Long
    private external fun nativeGenerate(
        contextHandle: Long,
        prompt: String,
        maxTokens: Int,
        temperature: Float,
        topP: Float,
        repeatPenalty: Float,
        tokenCallback: (String) -> Boolean
    )
    private external fun nativeFreeContext(contextHandle: Long)
    private external fun nativeFreeModel(modelHandle: Long)
}
